File: knapsack.java
// 2021 UCF HSPT - Knapsack?
// Author: Natalie Longtin

/*
the traditional knapsack problem can be solved with a dynamic programming
algorithm in O(nm) time, where n is the number of items and m is the capacity of the knapsack.
for this problem, however, this approach is too slow due to the large bounds on m.

consider reversing the problem: start by taking every single object, and then run
knapsack to find the minimum value of items that we must remove to get below the allowed capacity.
taking into account the unusual bounds, observe that we will only need to consider
removing items with a total weight of 55,000 or less.
*/

import java.util.Arrays;
import java.util.Scanner;

public class knapsack {
	public static void main(String[] args) {
		Scanner scan=new Scanner(System.in);
		
		//process test cases
		int t=scan.nextInt();
		for(int tt=0;tt<t;tt++) {
			n=scan.nextInt();
			m=scan.nextInt();
			
			values=new int[n];
			weights=new int[n];
			
			int totalweight=0;
			int totalvalue=0;
			for(int i=0;i<n;i++) {
				values[i]=scan.nextInt();
				weights[i]=scan.nextInt();
				totalvalue+=values[i];
				totalweight+=weights[i];
			}
			
			//if the total weight of all objects is less than the max capacity,
			//we can just take them all, our job is already done:)
			if(totalweight<=m) {
				System.out.println(totalvalue);
				continue;
			}
			
			overcapacity=totalweight-m;
			
			//otherwise, we are totalweight-m over capacity.
			//we must figure out some subset of items to leave behind,
			//such that this subset of items has a total weight greater than totalweight-m,
			//and the total value of this subset of items is minimal.

			dp=new int[n][overcapacity+1];
			for(int[] d:dp) Arrays.fill(d,-1);
			
			//at last, we need to subtract this minimum value to remove
			//from the total value of all the items.
			System.out.println(totalvalue-go(0,0));
		}
	}
	public static int go(int at, int rem) {
		if(at==n) {
			//we have reached the end of the array.
			//if we have removed enough items, this is a valid way to go
			//otherwise, we haven't reached the goal and this is an approach
			//that can never be considered; we return an arbitrarily high value to be sure
			
			return rem==overcapacity?0:INF;
		}
		
		if(dp[at][rem]!=-1) {
			//we have already calculated the answer for this state
			//and stored the answer in this array
			return dp[at][rem];
		}
		
		//res = the min sum of values of the items we must remove from this point on
		int res=INF;
		
		//we do not remove this item from the knapsack; we simply move on
		res=Math.min(res,go(at+1,rem));
		
		//we choose to remove this item; the weight we removed increases,
		//and the sum of values we removed also increases
		res=Math.min(res,values[at]+go(at+1,Math.min(overcapacity,rem+weights[at])));
		
		return dp[at][rem]=res;
	}
	static int n,m;
	static int overcapacity;

	static int[] values, weights;
	static int[][] dp;
	
	static final int INF=Integer.MAX_VALUE/2-5;
}

File: knapsack.cpp
/*

The bounds don't allow for the traditional knapsack solution which requires a DP state of n*m.
Instead, we are going to assume we can take every item (get the sum of all values and weights).
Then, we check how much over m we are (totWeight - m). This is how much we need to drop.
Due to the bounds of the problem, this is at most 55,000. This means our state is 1,000*55,000 at most.
Then, we do a take-it-or-leave-it DP that minimizes the value of the items we need to drop.

*/

#include <bits/stdc++.h>
using namespace std;

int n;
int BIG_NUMBER = INT_MAX / 4;
vector<int> values, weights;
vector<vector<int>> dp;

// This is our DP solution
int go(int indx, int currW) {
    // If we drop enough weight at any time, we can stop dropping things
    if (currW <= 0) return 0;
    // If we reach the end of our array and haven't dropped enough weight, this is not a valid solution
    // Return a big number to ensure this path will not be our solution
    if (indx >= n) return BIG_NUMBER;
    // If we already saw this state, return the answer we got
    if (dp[indx][currW] != -1) return dp[indx][currW];

    // If we drop this item, we lose values[indx] and need to update our current weight
    int drop = values[indx] + go(indx + 1, currW - weights[indx]);
    // If we keep this item, we go to the next one
    int keep = go(indx + 1, currW);

    // Our value at this state is the min of drop and keep
    return dp[indx][currW] = min(drop, keep);
}

// Per test case solution
void solve() {
    // Input
    cin >> n;
    int m;
    cin >> m;

    values.assign(n, 0);
    weights.assign(n, 0);
    int totValue = 0, totWeight = 0;
    for (int i = 0; i < n; i++)
    {
        cin >> values[i];
        cin >> weights[i];
        totValue += values[i];      // Sum of all values
        totWeight += weights[i];    // Sum of all weights
    }

    // If we can take it all and not exceed capacity, let's do it
    if (totWeight <= m) cout << totValue << endl;
    // Otherwise, we need to do a take-it-or-leave-it DP
    else {
        // This is how much we need to drop
        int newM = totWeight - m;
        // Fill our DP table with -1
        dp.assign(n, vector<int>(newM+1, -1));
        // Do the DP
        cout << (totValue - go(0, newM)) << endl;
    }
}

int main() {
    // Number of input cases
    int t;
    cin >> t;

    while (t-- > 0)
        solve();

    return 0;
}